File: Python/programmers/etc/level3/섬_연결하기.py
def solution(n, costs):
    
    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]
    
    def union(a, b):
        a, b = find(a), find(b)
        parent[max(a, b)] = min(a, b)
    
    answer = 0
    cnt = 0
    parent = list(range(n))
    costs.sort(key=lambda t: t[2])
    
    for a, b, cost in costs:
        if find(a) != find(b):
            answer += cost
            union(a, b)
            cnt += 1
        if cnt == n-1:
            break
    
    return answer


n = 4
costs = [[0,1,1],[0,2,2],[1,2,5],[1,3,1],[2,3,8]]
print(solution(n, costs)) # 4

File: Java/programmers/etc/level3/섬연결하기.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class 섬연결하기 {
    public int solution(int n, int[][] costs) {
        int answer = 0;
        int cnt = 0;
        
        int[] parent = new int[n];
        for (int i=0; i<n; i++) {
            parent[i] = i;
        }
        Arrays.sort(costs, (o1,o2) -> o1[2]-o2[2]);

        for (int[] c : costs) {
            int a = c[0], b = c[1], cost = c[2];
            if (find(parent, a) != find(parent, b)) {
                answer += cost;
                union(parent, a, b);
                cnt++;
            }
            if (cnt == n-1) break;
        }

        return answer;
    }

    private static int find(int[] parent, int x) {
        if (parent[x] != x) {
            parent[x] = find(parent, parent[x]);
        }
        return parent[x];
    }

    private static void union(int[] parent, int a, int b) {
        int pa = find(parent, a), pb = find(parent, b);
        parent[Math.max(pa,pb)] = Math.min(pa,pb);
    }
}
